pkg/controller/job/job_controller.go

```diff
@@ -808,15 +808,20 @@ func (jm *Controller) syncJob(ctx context.Context, key string) (rErr error) {
 	var manageJobErr error
 
 	exceedsBackoffLimit := jobCtx.failed > *job.Spec.BackoffLimit
+	successCriteriaMetCond := successCriteriaMetCondition(&job)
 
-	if feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) {
+	// Transition of successCriteriaMet to Failed or FailureTarget is not allowed.
+	if successCriteriaMetCond == nil && feature.DefaultFeatureGate.Enabled(features.JobPodFailurePolicy) {
 		if failureTargetCondition := findConditionByType(job.Status.Conditions, batch.JobFailureTarget); failureTargetCondition != nil {
 			jobCtx.finishedCondition = newFailedConditionForFailureTarget(failureTargetCondition, jm.clock.Now())
 		} else if failJobMessage := getFailJobMessage(&job, pods); failJobMessage != nil {
 			// Prepare the interim FailureTarget condition to record the failure message before the finalizers (allowing removal of the pods) are removed.
 			jobCtx.finishedCondition = newCondition(batch.JobFailureTarget, v1.ConditionTrue, batch.JobReasonPodFailurePolicy, *failJobMessage, jm.clock.Now())
 		}
 	}
+	if successCriteriaMetCond != nil {
+		jobCtx.finishedCondition = successCriteriaMetCond
```

Comment:
Hm, I'm thinking if it was maybe cleaner to have `jobCtx.successCondition`, did you consider that? I'm not sure about the impact on code overall, but it seem reasonable alternative. The current one we could rename to `jobCtx.failureCondiiton`.

Reply:
I also feel that the current approach is a bit hacky.
As I can see in the job controller code, we can easily have 2 conditions for failure and success, easily.

Let me select the approach having `successCondition`, then let me hear what you think about the new logic.

Reply:
Done.